File: C++/2-TwoPointers/Algorithms/1-Fast-and-SlowPointers(Floyd’sCycleDetection)/Fast-and-SlowPointers.cpp
#include <iostream>
#include <unordered_set>
#include <vector>

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

// Helper function to create a linked list from a vector
ListNode* createLinkedList(const std::vector<int>& values, int cyclePos = -1) {
    ListNode* head = nullptr;
    ListNode* tail = nullptr;
    ListNode* cycleNode = nullptr;

    for (int i = 0; i < values.size(); ++i) {
        ListNode* newNode = new ListNode(values[i]);
        if (!head) {
            head = newNode;
            tail = newNode;
        } else {
            tail->next = newNode;
            tail = newNode;
        }
        if (i == cyclePos) {
            cycleNode = newNode;
        }
    }
    if (cyclePos != -1) {
        tail->next = cycleNode; // Create the cycle
    }
    return head;
}

// Helper function to print the linked list
void printLinkedList(ListNode* head) {
    ListNode* current = head;
    std::unordered_set<ListNode*> visited;
    while (current) {
        if (visited.find(current) != visited.end()) {
            std::cout << "Cycle detected, stopping here." << std::endl;
            break;
        }
        visited.insert(current);
        std::cout << current->val << " -> ";
        current = current->next;
    }
    if (current == nullptr)
        std::cout << "nullptr" << std::endl;
}

// Helper function to free the linked list
void freeLinkedList(ListNode* head) {
    ListNode* current = head;
    while (current) {
        ListNode* next = current->next;
        delete current;
        current = next;
    }
}

// 1. Using a hash set to detect cycles.
//   - Time Complexity: O(n), where n is the number of nodes in the list.
//   - Space Complexity: O(n), for the hash set.
bool hasCycleUsingHashSet(ListNode *head) {
    std::unordered_set<ListNode*> visitedNodes;
    ListNode* current = head;
    while (current) {
        if (visitedNodes.find(current) != visitedNodes.end()) {
            return true; // Cycle detected: current node is already in the set.
        }
        visitedNodes.insert(current); // Add the current node to the set.
        current = current->next;
    }
    return false; // No cycle found: reached the end of the list.
}

// 2. Floyd's Cycle Detection Algorithm (Fast and Slow Pointers).
//   - Time Complexity: O(n), where n is the number of nodes in the list.
//   - Space Complexity: O(1), constant extra space.
bool hasCycleFloyd(ListNode *head) {
    if (!head || !head->next) {
        return false; // No cycle if the list is empty or has only one node.
    }
    ListNode* slow = head;
    ListNode* fast = head->next; // Start fast pointer one step ahead.

    while (fast && fast->next) {
        if (slow == fast) {
            return true; // Cycle detected: fast and slow pointers meet.
        }
        slow = slow->next;         // Slow pointer moves one step.
        fast = fast->next->next; // Fast pointer moves two steps.
    }
    return false; // No cycle found: fast pointer reached the end of the list.
}

// 3. Modifying the list nodes (Not recommended for read-only lists).
//   - Time Complexity: O(n), where n is the number of nodes in the list.
//   - Space Complexity: O(1), constant extra space.
bool hasCycleModifyingList(ListNode *head) {
    ListNode* current = head;
    ListNode* temp = new ListNode(0); // Use a dummy node as a marker.
    while (current) {
        if (current->next == temp) {
            delete temp;
            return true; // Cycle detected: next pointer points to the marker.
        }
        ListNode* next = current->next;
        current->next = temp; // Change next pointer to the marker.
        current = next;
    }
    delete temp;
    return false; // No cycle found: reached the end of the list.
}

// 4. Using a vector to store visited nodes (Similar to HashSet, but uses vector).
//   - Time Complexity: O(n^2) in the worst case, where n is the number of nodes.
//   - Space Complexity: O(n), for the vector.
bool hasCycleUsingVector(ListNode *head) {
    std::vector<ListNode*> visitedNodes;
    ListNode* current = head;
    while (current) {
        for (ListNode* node : visitedNodes) {
            if (node == current) {
                return true; // Cycle detected: current node is in the vector.
            }
        }
        visitedNodes.push_back(current); // Add the current node to the vector.
        current = current->next;
    }
    return false; // No cycle found: reached the end of the list.
}

// 5.  Check for cycle by comparing the address of the nodes.
//    - Time complexity: O(n^2)
//    - Space complexity: O(1)
bool hasCycleCompareAddress(ListNode* head) {
    if (!head) return false;

    ListNode* current = head;
    while (current) {
        ListNode* runner = head;
        while (runner != current) {
            if (runner == nullptr)
                break;
            if (runner->next == current)
                return true;
            runner = runner->next;
        }
        current = current->next;
    }
    return false;
}

int main() {
    // Test cases with and without cycles.
    std::vector<int> list1 = {1, 2, 3, 4, 5};
    std::vector<int> list2 = {1, 2, 3, 4, 5};
    std::vector<int> list3 = {1};
    std::vector<int> list4 = {};
    std::vector<int> list5 = {1, 2};

    ListNode* head1 = createLinkedList(list1);             // No cycle
    ListNode* head2 = createLinkedList(list2, 2);    // Cycle at index 2 (3)
    ListNode* head3 = createLinkedList(list3, 0);    // Cycle at index 0
    ListNode* head4 = createLinkedList(list4);           // Empty List
    ListNode* head5 = createLinkedList(list5, 1);

    std::cout << "Test Case 1: No Cycle" << std::endl;
    std::cout << "List: ";
    printLinkedList(head1);
    std::cout << "HashSet: " << (hasCycleUsingHashSet(head1) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Floyd's: " << (hasCycleFloyd(head1) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Modifying List: " << (hasCycleModifyingList(head1) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Vector: " << (hasCycleUsingVector(head1) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Compare Address: " << (hasCycleCompareAddress(head1) ? "Cycle Detected" : "No Cycle") << std::endl;
    freeLinkedList(head1);

    std::cout << "\nTest Case 2: Cycle at index 2" << std::endl;
    std::cout << "List: ";
    printLinkedList(head2);
    std::cout << "HashSet: " << (hasCycleUsingHashSet(head2) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Floyd's: " << (hasCycleFloyd(head2) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Modifying List: " << (hasCycleModifyingList(head2) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Vector: " << (hasCycleUsingVector(head2) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Compare Address: " << (hasCycleCompareAddress(head2) ? "Cycle Detected" : "No Cycle") << std::endl;
    freeLinkedList(head2);

    std::cout << "\nTest Case 3: Cycle at head" << std::endl;
    std::cout << "List: ";
    printLinkedList(head3);
    std::cout << "HashSet: " << (hasCycleUsingHashSet(head3) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Floyd's: " << (hasCycleFloyd(head3) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Modifying List: " << (hasCycleModifyingList(head3) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Vector: " << (hasCycleUsingVector(head3) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Compare Address: " << (hasCycleCompareAddress(head3) ? "Cycle Detected" : "No Cycle") << std::endl;
    freeLinkedList(head3);

    std::cout << "\nTest Case 4: Empty List" << std::endl;
    std::cout << "List: ";
    printLinkedList(head4);
    std::cout << "HashSet: " << (hasCycleUsingHashSet(head4) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Floyd's: " << (hasCycleFloyd(head4) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Modifying List: " << (hasCycleModifyingList(head4) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Vector: " << (hasCycleUsingVector(head4) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Compare Address: " << (hasCycleCompareAddress(head4) ? "Cycle Detected" : "No Cycle") << std::endl;
    freeLinkedList(head4);

    std::cout << "\nTest Case 5: Cycle at tail" << std::endl;
    std::cout << "List: ";
    printLinkedList(head5);
    std::cout << "HashSet: " << (hasCycleUsingHashSet(head5) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Floyd's: " << (hasCycleFloyd(head5) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Modifying List: " << (hasCycleModifyingList(head5) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Vector: " << (hasCycleUsingVector(head5) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Compare Address: " << (hasCycleCompareAddress(head5) ? "Cycle Detected" : "No Cycle") << std::endl;
    freeLinkedList(head5);

    return 0;
}

File: Java/2-TwoPointers/Algorithms/1-Fast-and-SlowPointers(Floyd’sCycleDetection)/Fast-and-SlowPointers.java
import java.util.*;

public class FastAndSlowPointers {

    // Definition of ListNode (used for linked list problems)
    static class ListNode {
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
            this.next = null;
        }
    }

    // Definition of Node (used for graph problems)
    static class Node {
        int val;
        List<Node> neighbors;

        Node(int val) {
            this.val = val;
            this.neighbors = new ArrayList<>();
        }
    }

    public static void main(String[] args) {
        // Example Usage for Linked List Cycle Detection
        ListNode head = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        ListNode node5 = new ListNode(5);

        head.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        node5.next = node2; // Creating a cycle: 5 points back to 2

        System.out.println("Linked List Cycle Detection:");
        System.out.println("Approach 1 (Floyd's Cycle Detection): Cycle detected? " + hasCycleFloyd(head)); // true
        System.out.println("Approach 2 (Using a Set): Cycle detected? " + hasCycleSet(head)); // true
        System.out.println("Approach 3 (Modified Floyd's): Cycle detected? " + hasCycleModifiedFloyd(head)); //true

        //Find the start of the cycle
        System.out.println("Start of cycle (Floyd's): " + findCycleStart(head).val);

        // Example Usage for Happy Number
        int num1 = 19;
        int num2 = 2;

        System.out.println("\nHappy Number:");
        System.out.println(num1 + " is a happy number? " + isHappy(num1)); // true
        System.out.println(num2 + " is a happy number? " + isHappy(num2)); // false

        // Example usage Palindrome Linked List
        ListNode headPalindrome1 = new ListNode(1);
        headPalindrome1.next = new ListNode(2);
        headPalindrome1.next.next = new ListNode(2);
        headPalindrome1.next.next.next = new ListNode(1);
        System.out.println("\nPalindrome Linked List:");
        System.out.println("Is the linked list a palindrome? " + isPalindrome(headPalindrome1));

        ListNode headPalindrome2 = new ListNode(1);
        headPalindrome2.next = new ListNode(2);
        System.out.println("Is the linked list a palindrome? " + isPalindrome(headPalindrome2));

    }

    // Approach 1: Floyd's Cycle Detection Algorithm (Fast and Slow Pointers)
    // - Time Complexity: O(n)
    // - Space Complexity: O(1)
    // - Explanation: Uses two pointers, 'slow' and 'fast'.  'Slow' moves one step at a time,
    //   'fast' moves two steps at a time. If there's a cycle, the two pointers will eventually meet.
    public static boolean hasCycleFloyd(ListNode head) {
        if (head == null || head.next == null) {
            return false; // No cycle if the list is empty or has only one node
        }

        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;      // Move slow pointer one step
            fast = fast.next.next;  // Move fast pointer two steps

            if (slow == fast) {
                return true; // Cycle detected: slow and fast pointers meet
            }
        }
        return false; // No cycle: fast pointer reached the end of the list
    }

    // Approach 2: Using a Set to detect cycles
    // - Time Complexity: O(n)
    // - Space Complexity: O(n)
    // - Explanation:  Iterate through the linked list, adding each node to a Set.
    //   If a node is encountered that's already in the Set, then there's a cycle.
    public static boolean hasCycleSet(ListNode head) {
        if (head == null) {
            return false;
        }
        Set<ListNode> visitedNodes = new HashSet<>();
        ListNode current = head;
        while (current != null) {
            if (visitedNodes.contains(current)) {
                return true; // Cycle detected: node already in the set
            }
            visitedNodes.add(current); // Add the current node to the set
            current = current.next;
        }
        return false; // No cycle: reached the end of the list
    }

    // Approach 3: Modified Floyd's Algorithm (Finding the cycle start)
    // - Time Complexity: O(n)
    // - Space Complexity: O(1)
    // - Explanation: This approach not only detects the cycle but also finds the start
    //   node of the cycle. After the slow and fast pointers meet, move the slow
    //   pointer back to the head and move both slow and fast pointers one step at a
    //   time. The point where they meet again is the start of the cycle.
    public static boolean hasCycleModifiedFloyd(ListNode head) {
        if (head == null || head.next == null) {
            return false;
        }

        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                return true; // Cycle detected
            }
        }
        return false;
    }

    public static ListNode findCycleStart(ListNode head) {
        if (head == null || head.next == null) {
            return null; // No cycle, no start
        }

        ListNode slow = head;
        ListNode fast = head;

        // Phase 1: Detect the cycle
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                break; // Cycle detected, exit the loop
            }
        }

        if (fast == null || fast.next == null) {
            return null; // No cycle found
        }

        // Phase 2: Find the start of the cycle
        slow = head; // Move slow pointer back to the head
        while (slow != fast) {
            slow = slow.next; // Move both pointers one step at a time
            fast = fast.next;
        }
        return slow; // The meeting point is the start of the cycle
    }

    // Approach 4: Happy Number Problem
    // - Time Complexity: O(log n) - The number of iterations is bounded by the number of digits in 'n'.
    // - Space Complexity: O(1)      - Constant extra space is used.
    // - Explanation: A happy number is a number that eventually reaches 1 when repeatedly replaced by the sum of the square of its digits.
    //   This solution uses the fast and slow pointer approach to detect if a number is happy or not.
    //   If a number is not happy, its sequence will end up in a cycle.  We use the fast and slow pointer
    //   technique to detect this cycle.
    public static boolean isHappy(int n) {
        int slow = n;
        int fast = n;

        do {
            slow = findSquareSum(slow);       // Move slow pointer one step
            fast = findSquareSum(findSquareSum(fast)); // Move fast pointer two steps
        } while (slow != fast);

        return slow == 1; // If the meeting point is 1, the number is happy
    }

    // Helper function to calculate the sum of the squares of the digits of a number
    private static int findSquareSum(int num) {
        int sum = 0;
        while (num > 0) {
            int digit = num % 10;
            sum += digit * digit;
            num /= 10;
        }
        return sum;
    }

    // Approach 5: Palindrome Linked List
    // - Time Complexity: O(n)
    // - Space Complexity: O(1)
    public static boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) {
            return true; // Empty list or single node is a palindrome
        }

        ListNode slow = head;
        ListNode fast = head;

        // Find the middle of the linked list
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // Reverse the second half of the linked list
        ListNode secondHalf = reverseList(slow);

        // Compare the first half and the reversed second half
        ListNode firstHalf = head;
        while (secondHalf != null) {
            if (firstHalf.val != secondHalf.val) {
                return false; // Not a palindrome
            }
            firstHalf = firstHalf.next;
            secondHalf = secondHalf.next;
        }
        return true; // Is a palindrome
    }

    // Helper function to reverse a linked list
    private static ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode current = head;
        ListNode next = null;

        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        return prev;
    }
}

